nixos/doc/manual/development/fallback-typing.xml

```diff
@@ -0,0 +1,135 @@
+  # More complicated transformations could take place as well
+  config.environment.etc."foo-config.json".text = builtins.toJSON config.services.foo.settings;
+
+}
+  </programlisting>
+  <para>
+   With such a module, users can now set arbitrary program settings, but only the <literal>port</literal> setting is checked against the more strict <literal>types.port</literal> type. An example of what's possible and what isn't:
+  </para>
+  <programlisting>
+services.foo.settings = {
+
+  # Even though this option isn't declared, this is allowed with fallback typing
+  logging_level = "DEBUG";
+
+  # Now allowed because there is no instances option, and the fallback type
+  # doesn't allow list values
+  # instances = [ "bort" ];
+
+  # Allowed since port is an option and this value matches its type
+  port = 80;
+
+  # Not allowed, since the port option is of the port type, which doesn't
+  # allow values over 65535
+  # port = 1000000;
+
+};
+  </programlisting>
+ </section>
+ <section xml:id="sec-fallback-typing-inf-rec">
+  <title>Infinite Recursion for Dependent Settings</title>
+  <para>
+   Commonly when <literal>types.attrsOf</literal> is used as an option type, infinite recursion occurs when some attribute depends on another attribute of this option. This happens because <literal>types.attrsOf</literal> is strict in values, meaning that even checking whether a certain attribute exists leads to all attribute <emphasis>values</emphasis> being evaluated.
+  </para>
+  <para>
+   The same occurs when <literal>types.attrsOf</literal> is used for fallback typing whenever a setting <emphasis>without</emphasis> a declared option is assigned to a value that depends on another setting. Since this doesn't happen for settings with a declared option, we can use this to our advantage to avoid infinite recursion.
+
+   So to avoid this problem: Whenever a setting <literal>foo</literal> depends on another setting <literal>bar</literal>, declare <literal>foo</literal> to be an option, such that it doesn't use the fallback type.
+
+   Here's an example:
+  </para>
+  <programlisting>
+{ lib, config, ... }: {
+
+  options.settings = lib.mkOption {
+    type = lib.types.submodule {
+
+      config._module.fallbackType = lib.types.attrsOf lib.types.str;
+
+      # Because foo depends on bar below, declare foo as an option to avoid
+      # infinite recursion
+      options.foo = lib.mkOption {
+        type = lib.types.str;
+        description = "The value for foo.";
+      };
+
+    };
+    default = {};
+  };
+
+  config.settings.bar = "bar";
+
+  # foo depends on bar here
+  config.settings.foo = config.settings.bar;
+
+}
+  </programlisting>
+  <note>
+   <para>
+    While the type <link linkend="types-lazyAttrsOf"><literal>types.lazyAttrsOf</literal></link> can be used to prevent infinite recursion as well, this has some downsides to it, and is not recommended unless the consequences are fully understood.
```

Comment:
```suggestion
    You could use the <link linkend="types-lazyAttrsOf"><literal>types.lazyAttrsOf</literal></link> to prevent infinite recursion, but it has only partial support for conditional definitions with <literal>mkIf</literal>.
```

No need to scare the reader.